Data Structures/matrix_sum_query.cpp

```diff
@@ -0,0 +1,84 @@
+// Name : Shivam Singhal
+// Email ID : shivamsinghal0610@gmail.com
+// Algorithm to calculate the sum of elements contained in a sub-matrix
+// formed by given 2 co-ordinates in a query taking first point as upper left
+// corner and the second point as the lower right corner
+
+#include <iostream>
+using namespace std;
+
+int main()
+{
+	int m, n, q;
+	cin >> m >> n;
+	cout << "Dimensions of matrix are : " << m <<", "<<n<<endl;
+	int a[m][n];
+	for(int i=0; i<m; i++){
+		for(int j=0; j<n; j++)
+			cin >> a[i][j];
+	}
+	cout << "Matrix elements are :" << endl;
+	for(int i=0; i<m; i++){
+		for(int j=0; j<n; j++)
+			cout << a[i][j] << " ";
+		cout << endl;
+	}
+	// taking column wise sum
+	for(int i=1; i<m; i++){
+		for(int j=0; j<n; j++){
+			a[i][j] += a[i-1][j];
+		}
+	}
+	// taking row wise sum 
+	for(int i=0; i<m; i++){
+		for(int j=1; j<n; j++){
+			a[i][j] += a[i][j-1];
+		}
+	}
+	// the sum matrix contains the sum contained in all elements
+	// from (0,0) to (i,j) at (i,j)
+	cout<<"Sum matrix:\n";
+	for(int i=0;i<m;i++){
+		for(int j=0;j<n;j++)
+			cout<<a[i][j]<<" ";
+		cout<<endl;
+	}
+	cin >> q;
+	cout << "Please enter the 2 co-ordinates for which sum is to be found.\n";
+	cout << "Make sure that the co-ordinates are 0-indexed.\n";
+	cout << "While the first co-ordinate should correspond to the top left co-ordinate, the second co-ordinate should correspond to the bottom right co-ordinate." << endl;
+	cout << "Number of queries : " << q << endl;
+	for(int i=1; i<=q; i++){
+		int ai, aj, bi, bj;
+		cin >> ai >> aj >> bi >> bj;
+		cout << "Query #" << i << " : " << "(" << ai << "," << aj << ") to (" << bi << "," << bj << ")" << endl;
+		// in case of improper rectangle
+		if	(ai > bi || aj > bj)
```

Comment:
Also, add the conditions when the coordinates entered are outside the range of the matrix.